### ISSUE TITLE
compileAndLintWithContext doesn't always find all problems

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#5391]
## Expected Behavior
```
WARN: Attempt to load key 'platform.random.idempotence.check.rate' for not yet loaded registry

java.lang.AssertionError: 
Expected size: 0 but was: 3 in:
[
    CodeSmell(issue=Issue(id='BooleanPropertyNaming', severity=CodeSmell, debt=5min), entity=Entity(name=uniqueFilms, signature=Test.kt$val uniqueFilms: Boolean = feed2.films.isEmpty(), location=Location(source=10:13, text=237:285, file=Test.kt, filePath=FilePath(absolutePath=Test.kt, basePath=null, relativePath=null)), ktElement=PROPERTY), message=Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueFilms, metrics=[], references=[], severity=WARNING, id='BooleanPropertyNaming'),
    CodeSmell(issue=Issue(id='BooleanPropertyNaming', severity=CodeSmell, debt=5min), entity=Entity(name=uniqueFilms2, signature=Test.kt$val uniqueFilms2: Boolean = feed2.films.map(Feed.Film::id).isEmpty(), location=Location(source=11:13, text=298:382, file=Test.kt, filePath=FilePath(absolutePath=Test.kt, basePath=null, relativePath=null)), ktElement=PROPERTY), message=Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueFilms2, metrics=[], references=[], severity=WARNING, id='BooleanPropertyNaming'),
    CodeSmell(issue=Issue(id='BooleanPropertyNaming', severity=CodeSmell, debt=5min), entity=Entity(name=uniqueAttributes, signature=Test.kt$val uniqueAttributes: Boolean = feed2.attributes.map(Feed.Attribute::code) .isEmpty(), location=Location(source=13:13, text=395:516, file=Test.kt, filePath=FilePath(absolutePath=Test.kt, basePath=null, relativePath=null)), ktElement=PROPERTY), message=Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueAttributes, metrics=[], references=[], severity=WARNING, id='BooleanPropertyNaming')
]

	at io.gitlab.arturbosch.detekt.rules.naming.BooleanPropertyNamingSpec.should warn about Kotlin Boolean(BooleanPropertyNamingSpec.kt:648)
```

## Observed Behavior
```
WARN: Attempt to load key 'platform.random.idempotence.check.rate' for not yet loaded registry

java.lang.AssertionError: 
Expected size: 0 but was: 1 in:
[CodeSmell(issue=Issue(id='BooleanPropertyNaming', severity=CodeSmell, debt=5min), entity=Entity(name=uniqueFilms, signature=Test.kt$val uniqueFilms = feed2.films.isEmpty(), location=Location(source=10:13, text=237:276, file=Test.kt, filePath=FilePath(absolutePath=Test.kt, basePath=null, relativePath=null)), ktElement=PROPERTY), message=Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueFilms, metrics=[], references=[], severity=WARNING, id='BooleanPropertyNaming')]

	at io.gitlab.arturbosch.detekt.rules.naming.BooleanPropertyNamingSpec.should warn about Kotlin Boolean(BooleanPropertyNamingSpec.kt:646)
```

## Steps to Reproduce
```kotlin
    @Test
    fun `should warn about Kotlin Boolean`() {
        val code = """
            class Feed(
                val films: List<Film>,
                val attributes: List<Attribute>,
            )
            class Film(val id: String)
            class Attribute(val code: String)
            fun f() {
                val feed2: Feed = generateSequence { Feed(emptyList(), emptyList()) }
                    .first { feed2 ->
                        val uniqueFilms = feed2.films.isEmpty()
                        val uniqueFilms2 =
                            feed2.films.map(Film::id).isEmpty()
                        val uniqueAttributes =
                            feed2.attributes.map(Attribute::code)
                                .isEmpty()
                        uniqueFilms && uniqueFilms2 && uniqueAttributes
                    }
            }
        """.trimIndent()
        val findings = subject.compileAndLintWithContext(env, code)

        assertThat(findings).hasSize(0) // explicitly require the wrong number (there should be 3)
    }
```

## Context
I was trying to write a repro for a problem I observed with `BooleanPropertyNaming`, and I noticed something strange with the output of `hasSize`. Note: I tried simplifying the repro but the problem disappeared. Probably worth mentioning that if I explicitly type the `val`s as `val ...: Boolean =` all 3 problems are flagged!

## Your Environment

* Version of detekt used: HEAD @ cdb59286 (latest `main` on 2022-10-08)
* Operating System and version: Windows

**Comment by schalkms** [comment]
The same happens with `lintWithContext()` considering we run this test snippet with JSR223 *disabled*. Hence, this has nothing to do with `compileAndLintWithContext()`.

If JSR223 is *enabled* everything is alright, since the **given Kotlin code is invalid**.

**Comment by TWiStErRob** [comment]
😒 my bad, I was relying too much on the injected code snippet highlight which didn't show red. Indeed, were some errors in the code, I have corrected them (edited OP), and the issue still stands as described in OP.

---

By JSR223 do you mean this?
```kotlin
private val shouldCompileTestSnippets: Boolean =
    System.getProperty("compile-snippet-tests", "false")!!.toBoolean()
```

I'm a bit confused by this toggle. If a Rule `@RequiresTypeResolution` like `BooleanPropertyNaming` does, why is it possible to run their tests without type resolution (i.e. compilation)? Or is the "type resolution" just the usage of BindingContext and compilation is not involved?

**Comment by schalkms** [comment]
[JSR223](https://www.openhab.org/docs/configuration/jsr223.html) is the scripting engine for JVM languages.
It's not related to `BindingContext`, since this API exists in the Kotlin compiler. Hence, you don't need to compile snippets with JSR223 in order to use `BindingContext` APIs for type resolution.

You found the right code snippet. You can either set this variable to true to compile all test snippets with JSR223 or use the Gradle parameter. Due to increased runtime this setting is false by default.

With the updated code snippet, 3 code smells are reported when `compileAndLintWithContext()` with `shouldCompileTestSnippets = true`. So this issue can be closed I think, unless there are more questions coming up.

**Comment by TWiStErRob** [comment]
Can be closed if is `shouldCompileTestSnippets=true` on CI, otherwise the testing framework leaves space for easy errors and we might have some false positive tests.

**Comment by TWiStErRob** [comment]
Sidenote: I really think that for type resolution tests, the values should be forcefully true, however slow.

**Comment by schalkms** [comment]
> Can be closed if is `shouldCompileTestSnippets=true` on CI, otherwise the testing framework leaves space for easy errors and we might have some false positive tests.

Yes, this is fulfilled with different CI checks including **compile-test-snippets**, which already does what you describe.

**Comment by TWiStErRob** [comment]
Any idea what we could do to improve the DX? Because if I run different Spec files randomly on my local machine I will either get a trustworthy test or a potential false positive. This false positive will be discovered on the CI, but that's a pretty big turnaround and headache time. I'm curious to look at

> I really think that for type resolution tests, the value should be forcefully true, however slow.

if you would accept that PR.

(Of course I can change the code or edit the run configuration but both of those are just hacks IMO, which have many unintended side effects.)

**Comment by BraisGabin** [comment]
You can find more background at #4874. I would be against compile the snippets by default. That's like adding tests to the tests. I can't recall any project that does that.

But I'm interested in ways to make this behavior more obvious: #4874
